pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -206,25 +211,37 @@ protected void setupInitialSegments(String realtimeTableName, ZNRecord partition
     List<ZNRecord> records = new ArrayList<>(nPartitions);
     final long now = System.currentTimeMillis();
     final int seqNum = 0; // Initial seq number for the segments
+
     for (int i = 0; i < nPartitions; i++) {
-      final List instances = partitionMap.get(Integer.toString(i));
-      LLCRealtimeSegmentZKMetadata metadata = new LLCRealtimeSegmentZKMetadata();
-      final String rawTableName = TableNameBuilder.extractRawTableName(realtimeTableName);
-      LLCSegmentName llcSegmentName = new LLCSegmentName(rawTableName, i, seqNum, now);
-      final String segName = llcSegmentName.getSegmentName();
-
-      metadata.setCreationTime(now);
-      metadata.setStartOffset(startOffset);
-      metadata.setNumReplicas(instances.size());
-      metadata.setTableName(rawTableName);
-      metadata.setSegmentName(segName);
-      metadata.setStatus(CommonConstants.Segment.Realtime.Status.IN_PROGRESS);
-
-      ZNRecord record = metadata.toZNRecord();
-      final String znodePath = PinotRealtimeSegmentManager.getSegmentsPath() + "/" + realtimeTableName + "/" + segName;
-      paths.add(znodePath);
-      records.add(record);
-      idealStateEntries.put(segName, instances);
+      SimpleConsumerWrapper kafkaConsumer = SimpleConsumerWrapper.forPartitionConsumption(new KafkaSimpleConsumerFactoryImpl(),
+          bootstrapHosts, "dummyClientId", topicName, i);
+      try {
+        final List instances = partitionMap.get(Integer.toString(i));
+        LLCRealtimeSegmentZKMetadata metadata = new LLCRealtimeSegmentZKMetadata();
+        final String rawTableName = TableNameBuilder.extractRawTableName(realtimeTableName);
+        LLCSegmentName llcSegmentName = new LLCSegmentName(rawTableName, i, seqNum, now);
+        final String segName = llcSegmentName.getSegmentName();
+
+        metadata.setCreationTime(now);
+
+        final long startOffset = kafkaConsumer.fetchPartitionOffset(initialOffset);
+        LOGGER.warn("Setting start offset for segment {} to {}", segName, startOffset);
+        metadata.setStartOffset(startOffset);
+        metadata.setEndOffset(Long.MAX_VALUE);
```

Comment:
`-1L` and `Long.MAX_VALUE` both mean no upper limit in `SimpleConsumerWrapper`. We actually don't check the end offset in `LLRealtimeSegmentDataManager` so technically this is a no-op, but if we were, it would mean that we would ignore all records with a offset larger or equal to zero.

If we were to update `LLRealtimeSegmentDataManager` to take into account the end offset, then having the ZK metadata already containing a very large end offset is better than having to add a special case for 0 later on.